packages/gatsby/lib/query-runner/query-compiler.js

```diff
@@ -0,0 +1,127 @@
+// @flow
+import path from 'path'
+import glob from 'glob'
+import Bluebird from 'bluebird'
+import invariant from 'invariant'
+import { IRTransforms } from 'relay-compiler'
+import ASTConvert from 'relay-compiler/lib/ASTConvert'
+import RelayPrinter from 'relay-compiler/lib/RelayPrinter'
+import RelayCompilerContext from 'relay-compiler/lib/RelayCompilerContext'
+import filterContextForNode from 'relay-compiler/lib/filterContextForNode'
+
+import { store } from '../redux'
+import FileParser from './file-parser'
+
+import type { DocumentNode, GraphQLSchema } from 'graphql'
+
+const { printTransforms } = IRTransforms
+const globp = Bluebird.promisify(glob)
+
+const {
+  ArgumentsOfCorrectTypeRule,
+  DefaultValuesOfCorrectTypeRule,
+  FragmentsOnCompositeTypesRule,
+  KnownTypeNamesRule,
+  LoneAnonymousOperationRule,
+  PossibleFragmentSpreadsRule,
+  ScalarLeafsRule,
+  VariablesAreInputTypesRule,
+  VariablesInAllowedPositionRule,
+} = require(`graphql`)
+
+type RootQuery = {
+  name: string,
+  path: string,
+  text: string,
+}
+
+class Runner {
+  baseDir: string
+  schema: GraphQLSchema
+
+  constructor(baseDir: string, schema: GraphQLSchema) {
+    this.baseDir = baseDir
+    this.schema = schema
+  }
+
+  async compileAll() {
+    let nodes = await this.parseEverything()
+    return this.write(nodes)
+  }
+
+  async parseEverything() {
+    let files = await globp(`${this.baseDir}/**/*.js`)
+    let parser = new FileParser()
+
+    return await parser.parseFiles(files)
+  }
+
+  async write(nodes: Map<string, DocumentNode>) {
+    const namePathMap = new Map()
+    const documents = []
+
+    nodes.forEach((doc, filePath) => {
+      documents.push(doc)
+      doc.definitions.forEach((def: any) => {
+        const name: string = def.name.value
+        namePathMap.set(name, filePath)
+      })
+    })
+
+    let compilerContext = new RelayCompilerContext(this.schema)
+    compilerContext = compilerContext.addAll(
+      ASTConvert.convertASTDocuments(this.schema, documents, [
+        ArgumentsOfCorrectTypeRule,
+        DefaultValuesOfCorrectTypeRule,
+        FragmentsOnCompositeTypesRule,
+        KnownTypeNamesRule,
+        LoneAnonymousOperationRule,
+        PossibleFragmentSpreadsRule,
+        ScalarLeafsRule,
+        VariablesAreInputTypesRule,
+        VariablesInAllowedPositionRule,
+      ])
+    )
+
+    const printContext = printTransforms.reduce(
+      (ctx, transform) => transform(ctx, this.schema),
+      compilerContext
+    )
+
+    const compiledNodes: Map<string, RootQuery> = new Map()
+
+    compilerContext.documents().forEach((node: { name: string }) => {
+      if (node.kind !== `Root`) return
+
+      const { name } = node
+      let filePath = namePathMap.get(name) || ``
+
+      invariant(
+        !compiledNodes.has(filePath),
+        `Gatsby: Pages may only specify one "root" query tag. ` +
+          `Combine them into a single query`
```

Comment:
I don't necessarily have a use-case, the single query per page is reasonable. Just wanted to note that you _could_ support more if needed. e.g you could have the exported name be the prop you inject.